File: src/tui/render/screen-frame.ts
import { center, composeTwoPane, stripAnsi, truncateAnsi } from "./text.js";

/**
 * The shared TUI screen layout, hoisted from the dashboard so every screen composes the same
 * way: a centered content block, a scrolling main column with ▲/▼ affordances, an optional
 * right details pane separated by whitespace (no vertical divider), and a full-width footer.
 * The dashboard and all subscreens render through this so their chrome stays uniform.
 */
export interface ScreenFrameInput {
  cols: number;
  rows: number;
  /** Header block (leading blank, title, full-width rule, trailing blank). */
  header: string[];
  /** Main column lines (2-space gutter, not pre-centered). */
  content: string[];
  /** Pre-styled footer hint lines (no rule, no indent); the frame adds both. */
  footerLines: string[];
  /** Index into `content` to keep visible; -1 to disable auto-scroll. */
  focusLine: number;
  /** Incoming persisted scroll offset (dashboard); subscreens pass 0/omit. */
  scrollOffset?: number;
  /** Show the right details pane when it both fits and the caller wants it. */
  twoPane?: boolean;
  /** Builds the right pane to an exact width/height; only called when twoPane. */
  rightPanel?: (panelWidth: number, height: number) => string[];
}

export interface ScreenFrameResult {
  frame: string;
  scrollOffset: number;
}

// Block geometry shared with the dashboard: a 72-floor centered block, 58% left column.
export function screenContentWidth(cols: number): number {
  return Math.max(72, cols);
}
export function screenLeftWidth(cols: number): number {
  return Math.max(32, Math.floor(screenContentWidth(cols) * 0.58));
}

export function composeScreenFrame(input: ScreenFrameInput): ScreenFrameResult {
  const { cols, rows, header, content } = input;
  const contentWidth = screenContentWidth(cols);
  const leftWidth = screenLeftWidth(cols);
  const centerInBlock = (line: string): string => truncateAnsi(center(line, contentWidth), cols);

  const footerIndent = "  ";
  const footer: string[] = [
    "─".repeat(Math.max(0, cols)),
    ...input.footerLines.map((line) => truncateAnsi(`${footerIndent}${line}`, cols)),
  ];

  const viewportHeight = Math.max(1, rows - header.length - footer.length);
  let scrollOffset = input.scrollOffset ?? 0;
  const focusLine = input.focusLine;
  // A focused card spans from its marker to the next blank line; reveal its whole body.
  let focusEnd = focusLine;
  while (focusEnd >= 0 && focusEnd + 1 < content.length && stripAnsi(content[focusEnd + 1] ?? "").trim() !== "") {
    focusEnd++;
  }
  const maxScroll = Math.max(0, content.length - viewportHeight);
  if (focusLine >= 0) {
    if (focusLine < scrollOffset + 1) {
      scrollOffset = Math.max(0, focusLine - 1);
    } else if (focusEnd >= scrollOffset + viewportHeight - 1) {
      scrollOffset = Math.min(maxScroll, focusEnd - viewportHeight + 2);
      if (focusLine < scrollOffset + 1) scrollOffset = Math.max(0, focusLine - 1);
    }
  }
  scrollOffset = Math.min(scrollOffset, maxScroll);

  const visible = content.slice(scrollOffset, scrollOffset + viewportHeight);
  const canScrollUp = scrollOffset > 0;
  const canScrollDown = scrollOffset < maxScroll;
  if (canScrollUp && visible.length > 0) visible[0] = centerInBlock("\x1b[2m▲ more ▲\x1b[0m");
  if (canScrollDown && visible.length > 0) visible[visible.length - 1] = centerInBlock("\x1b[2m▼ more ▼\x1b[0m");
  while (visible.length < viewportHeight) visible.push("");

  let body = visible;
  if (input.twoPane && input.rightPanel) {
    const panelWidth = Math.max(20, contentWidth - leftWidth - 4);
    const rightPanel = input.rightPanel(panelWidth, viewportHeight);
    // Whitespace separator (no vertical divider), matching the dashboard.
    body = composeTwoPane(visible, rightPanel, contentWidth, "   ");
  }

  return {
    frame: "\x1b[2J\x1b[H" + [...header, ...body, ...footer].join("\r\n"),
    scrollOffset,
  };
}

File: native/crates/aimux/src/tui_render/screen_frame.rs
use super::text::{center, compose_two_pane, strip_ansi, truncate_ansi};

#[derive(Debug, Clone)]
pub struct ScreenFrameInput<'a> {
    pub cols: usize,
    pub rows: usize,
    pub header: &'a [String],
    pub content: &'a [String],
    pub footer_lines: &'a [String],
    pub focus_line: isize,
    pub scroll_offset: usize,
    pub two_pane: bool,
    pub right_panel: Option<&'a [String]>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ScreenFrameResult {
    pub frame: String,
    pub scroll_offset: usize,
}

pub fn screen_content_width(cols: usize) -> usize {
    72.max(cols)
}

pub fn screen_left_width(cols: usize) -> usize {
    32.max(((screen_content_width(cols) as f64) * 0.58).floor() as usize)
}

pub fn compose_screen_frame(input: &ScreenFrameInput<'_>) -> ScreenFrameResult {
    let cols = input.cols;
    let content_width = screen_content_width(cols);
    let center_in_block = |line: &str| truncate_ansi(&center(line, content_width), cols);
    let footer_indent = "  ";
    let mut footer = vec!["─".repeat(cols)];
    footer.extend(
        input
            .footer_lines
            .iter()
            .map(|line| truncate_ansi(&format!("{footer_indent}{line}"), cols)),
    );

    let viewport_height = 1.max(input.rows.saturating_sub(input.header.len() + footer.len()));
    let mut scroll_offset = input.scroll_offset;
    let focus_line = input.focus_line;
    let mut focus_end = focus_line;
    while focus_end >= 0 {
        let next = (focus_end + 1) as usize;
        if next >= input.content.len() || strip_ansi(&input.content[next]).trim().is_empty() {
            break;
        }
        focus_end += 1;
    }
    let max_scroll = input.content.len().saturating_sub(viewport_height);
    if focus_line >= 0 {
        let focus = focus_line as usize;
        let end = focus_end.max(focus_line) as usize;
        if focus < scroll_offset + 1 {
            scroll_offset = focus.saturating_sub(1);
        } else if end >= scroll_offset + viewport_height.saturating_sub(1) {
            scroll_offset = max_scroll.min(end.saturating_sub(viewport_height).saturating_add(2));
            if focus < scroll_offset + 1 {
                scroll_offset = focus.saturating_sub(1);
            }
        }
    }
    scroll_offset = scroll_offset.min(max_scroll);

    let mut visible = input
        .content
        .iter()
        .skip(scroll_offset)
        .take(viewport_height)
        .cloned()
        .collect::<Vec<_>>();
    let can_scroll_up = scroll_offset > 0;
    let can_scroll_down = scroll_offset < max_scroll;
    if can_scroll_up && !visible.is_empty() {
        visible[0] = center_in_block("\x1b[2m▲ more ▲\x1b[0m");
    }
    if can_scroll_down && !visible.is_empty() {
        let last = visible.len() - 1;
        visible[last] = center_in_block("\x1b[2m▼ more ▼\x1b[0m");
    }
    while visible.len() < viewport_height {
        visible.push(String::new());
    }

    let body = if input.two_pane {
        if let Some(right_panel) = input.right_panel {
            compose_two_pane(&visible, right_panel, content_width, Some("   "))
        } else {
            visible
        }
    } else {
        visible
    };

    ScreenFrameResult {
        frame: format!(
            "\x1b[2J\x1b[H{}",
            input
                .header
                .iter()
                .chain(body.iter())
                .chain(footer.iter())
                .cloned()
                .collect::<Vec<_>>()
                .join("\r\n")
        ),
        scroll_offset,
    }
}
